hibernate-core/src/test/java/org/hibernate/orm/test/mapping/lazytoone/onetoone/JoinFetchedOneToOneAllowProxyTests.java

```diff
@@ -0,0 +1,197 @@
+import org.hibernate.testing.bytecode.enhancement.EnhancementOptions;
+import org.hibernate.testing.jdbc.SQLStatementInterceptor;
+import org.hibernate.testing.junit4.BaseNonConfigCoreFunctionalTestCase;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import static javax.persistence.FetchType.LAZY;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hibernate.annotations.FetchMode.JOIN;
+
+/**
+ * @author Steve Ebersole
+ */
+@RunWith( BytecodeEnhancerRunner.class)
+@EnhancementOptions( lazyLoading = true )
+public class JoinFetchedOneToOneAllowProxyTests extends BaseNonConfigCoreFunctionalTestCase {
+	private SQLStatementInterceptor sqlStatementInterceptor;
+
+	@Override
+	protected void applyMetadataSources(MetadataSources sources) {
+		super.applyMetadataSources( sources );
+		sources.addAnnotatedClass( Customer.class );
+		sources.addAnnotatedClass( SupplementalInfo.class );
+	}
+
+	@Override
+	protected void configureStandardServiceRegistryBuilder(StandardServiceRegistryBuilder ssrb) {
+		super.configureStandardServiceRegistryBuilder( ssrb );
+		ssrb.applySetting( AvailableSettings.ALLOW_ENHANCEMENT_AS_PROXY, true );
+		sqlStatementInterceptor = new SQLStatementInterceptor( ssrb );
+	}
+
+	@Test
+	public void testOwnerIsProxy() {
+		sqlStatementInterceptor.clear();
+
+		final EntityPersister supplementalInfoDescriptor = sessionFactory().getMetamodel().entityPersister( SupplementalInfo.class );
+		final BytecodeEnhancementMetadata supplementalInfoEnhancementMetadata = supplementalInfoDescriptor.getBytecodeEnhancementMetadata();
+		assertThat( supplementalInfoEnhancementMetadata.isEnhancedForLazyLoading(), is( true ) );
+
+		final EntityPersister customerDescriptor = sessionFactory().getMetamodel().entityPersister( Customer.class );
+		final BytecodeEnhancementMetadata customerEnhancementMetadata = customerDescriptor.getBytecodeEnhancementMetadata();
+		assertThat( customerEnhancementMetadata.isEnhancedForLazyLoading(), is( true ) );
+
+		inTransaction(
+				(session) -> {
+					final SupplementalInfo supplementalInfo = session.byId( SupplementalInfo.class ).getReference( 1 );
+
+					// we should have just the uninitialized SupplementalInfo proxy
+					//		- therefore no SQL statements should have been executed
+					assertThat( sqlStatementInterceptor.getSqlQueries().size(), is( 0 ) );
+
+					// access the id - should do nothing with db
+					supplementalInfo.getId();
+					assertThat( sqlStatementInterceptor.getSqlQueries().size(), is( 0 ) );
+
+					// this should trigger loading the entity's base state which should include join fetching Customer
+					supplementalInfo.getSomething();
+					assertThat( sqlStatementInterceptor.getSqlQueries().size(), is( 1 ) );
+
+					// should not trigger a load and the `customer` reference should be an uninitialized enhanced proxy
```

Comment:
This comment is a bit confusing as the customer was already initialized through the access of the entity's base state. The combination `@Fetch( JOIN )` and `@ManyToOne( fetch = LAZY )` seems odd to me. Would `session.byId( SupplementalInfo.class ).load( 1 )` also cause a join? If so, IMO this is "wrong".

Reply:
There is a discrepancy here between what should happen (imo) and what does happen.  What does happen (this is due to nothing I changed) is that this mapping chooses JOIN (eager) and ignores the LAZY.

This comes back to something I tried explaining to you and Andrea the other day.  What I think should happen here (although this would have to wait for 6) is that we'd simply not select the FK initially and later do a join when we access that association.

I think it is "odd" because of our implementation, not because the combo does not make sense

Reply:
>What does happen (this is due to nothing I changed) is that this mapping chooses JOIN (eager) and ignores the LAZY.

That's what I suspected. Just as follow-up, are we logging a warning for this (LAZY is ignored due to `@Fetch( JOIN )` already, or do you think this is not necessary?

>This comes back to something I tried explaining to you and Andrea the other day. What I think should happen here (although this would have to wait for 6) is that we'd simply not select the FK initially and later do a join when we access that association.

I'm a bit unsure what the effect of `@Fetch( JOIN )` should be for `fetch = LAZY` associations. In 5.x we apparently mark the association as `EAGER`. Do I understand that correctly, that you would like to honor `fetch = LAZY` in 6.0 by deferring the fetching of the FK? We can continue the discussion elsewhere if you like.

Reply:
> > What does happen (this is due to nothing I changed) is that this mapping chooses JOIN (eager) and ignores the LAZY.
> 
> That's what I suspected. Just as follow-up, are we logging a warning for this (LAZY is ignored due to `@Fetch( JOIN )` already, or do you think this is not necessary?

We do not as far as I know.  Not sure it is super important.  In general there are many combinations of annotations where one is ignored because of others that we do not log.  

Also, here, there is the whole "this will (likely) change later" aspect.


> I'm a bit unsure what the effect of `@Fetch( JOIN )` should be for `fetch = LAZY` associations. In 5.x we apparently mark the association as `EAGER`. Do I understand that correctly, that you would like to honor `fetch = LAZY` in 6.0 by deferring the fetching of the FK? We can continue the discussion elsewhere if you like.

Yeah, I don't think this is the place to discuss that.

But in general, its more than delaying fetching the FK.  Yes we would not select the FK but when we go to initialize it we would do the join.  E.g.

```
@Entity
class Address {
    ... 
    @ManyToOne( ..., fetch=LAZY )
    @Fetch( JOIN )
    PostalCode postalCode;
}

@Entity
class PostalCode {
    ...
}

Session session = ...;

// Loads the Address "base state", which does not include `postalCode`
//     select a.id, ... from address a where a.id = 1
Address addr = session.get( Address.class, 1 );

// Access Address#postalCode which triggers initializing PostalCode
//     select a.postal_code_id, p.id, p... from address a join postal_code p on ... where a.id = 1
```

This gets into a more in-depth discussion regarding how we can specify when FKs get selected.  To me, the first level is `@LazyGroup`.  This example mapping almost makes me think we should have an "implicit" group for any many-to-one with `FetchMode#JOIN`.

But again, imo these are details outside the scope of this PR